submissions/YuraZagor/tjsw/index.js

```diff
@@ -0,0 +1,157 @@
+const manNamesArr = ['Vasil', 'Yurij', 'Zahar', 'Taras', 'Olexiy', 'Oleg', 'Mykola', 'Sashko', 'Sergij', 'Bohdan']
+const womanNamesArr = ['Hanna', 'Helena', 'Vasilisa', 'Yulya', 'Oksana', 'Olena', 'Olga', 'Myroslava', 'Svitlana', 'Bohdana']
+const animalNamesArr = ['Joy', 'Happy', 'Lucky', 'Puffy']
+
+let catSay = 'meau';
+let dogSay = 'woufff';
+let humanSay = 'Glory to Ukraine!';
+
+function randomSelect(arr){
+  return arr[Math.floor(Math.random()*arr.length)]
+}
+let manName = randomSelect(manNamesArr)
+let womanName = randomSelect(womanNamesArr)
+let animalName = randomSelect(animalNamesArr)
+
+const inhabitants = [
+  { species: 'human', 
+    gender: 'male',
+    name: 'Vasil',
+    saying: 'Slava natsiyi',
+    friend: 'Hanna',
+    legs: 2,
+    hands: 2,
+  },
+  { species: 'human', 
+    gender: 'female',
+    name: 'Hanna',
+    saying: 'Smert vorogam',
+    friend: 'Vasil',
+    legs: 2,
+    hands: 2,
+  },
+];
+
+class Inhabitant {  
+  constructor(species, name, saying, friend, gender){
+    this.species=species;
+    this.name=name;
+    this.saying = saying
+    this.friend=friend;
+    this.gender=gender;
+  }
+};
+
+class Animal extends Inhabitant{
+  constructor( species, name, saying, friend, gender, paws, tail ){
+    super(species, name, saying, friend, gender);
+    this.paws=paws;
+    this.tail=tail;
+  };
+  populate(){
+    const { species, name, saying, friend, gender, paws, tail} = this
+    const inhabitantProperties = [species, name, saying, friend, gender, paws, tail]
+    print(inhabitantProperties.join('; '),'div');
+    inhabitants.push(this);
```

Comment:
(1) Repetition of code (see lines 77-81) below. Use method overloading/overriding. Also each class should handle the properties it owns immediately. Use extended `super` API to call inherited method version.
(2) Separation of concerns: this method does at least two distinctive things.
(3) Why would class `Animal` (or any class in hierarchy of inhabitants) populate the world? Please do not implement "mating" mechanisms. Anyway, class shouldn't refer to a global variable (constants containing ytuly constant values are still an exception to this rule).

Reply:
1) I am rather weak in this topic, tried my best but not sure I got the idea the right way. Anyways, now it seems better structured, and still works
2) Got it, that really adds to readability and makes code DRYer
3) I believe I shall insist: If someone inhabits a place, he lives in it himself being an inhabitant; if I populate a place, I get other people to live in it and I do not need to live there myself. No mating implyed, it can be done with children, or older people, or  say men only (like Valaam island in USSR)

Reply:
> 3. I believe I shall insist: If someone inhabits a place, he lives in it himself being an inhabitant; if I populate a place, I get other people to live in it and I do not need to live there myself. No mating implied, it can be done with children, or older people, or  say men only (like Valaam island in USSR)

Requrement re global variables in class definitions mentioned in Frogger task is still valid. It is not specific to Frogger. It is about OOP. You may be weak in any topic but you need to accumulate knowledge and re-use it as you go.

If you want to learn, ask questions and dive deeper into the topic (once again if needed).

Direct references to entities that exist outside class definitions restrict class code usage. Current implementation enforces peer developers to have a single one container for inhabitants and it has to be named `inhabitants`. OOP helps to be flexible and introduce concepts with minimal effect on code.

Better design would allow us to add containers like cities, communities etc - all populated with the same inhabitants (in different combinations). Reiterate over OOP principles if needed and improve this part.

Code should be extensible and maintainable.

"It works" is not a good argument when we're talking about design. 
Imagine the task "we may have 1 to 4 containers, each containing equal number of bags of rice. Container's max capacity is 4. Write a function to calculate total number of bags". The following fragment also works.  doesn't mean it is good.
```
function countBags(containersCount, bagsPerContainer) {
  if (containersCount=== 1) return bagsPerContainer * 1;
  if (containersCount=== 2) return bagsPerContainer * 2;
  if (containersCount=== 3) return bagsPerContainer * 3;
  if (containersCount=== 4) return bagsPerContainer * 4;
}
```

Reply:
added passing a destination array for inhabitants instances.  

Also even I can see the repetitive pattern in my function creation, but I haven't come up with a solution yet